resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -988,66 +1037,94 @@ protected void doResolveAll(String inetHost,
+                    result = secondaryResult;
+                }
+            }
+            result.complete(promise);
+            return;
+        }
+
+        // step 2
+        result = doResolveAllCached(hostname, additionals, resolveCacheFallback, resolvedInternetProtocolFamilies);
+        if (result != null) {
+            result.complete(promise);
+            promise = executor().newPromise();
         }
+        doResolveAllUncached(hostname, additionals, promise, promise, resolveCache, resolveCacheFallback,
+                completeOncePreferredResolved);
     }
 
-    static boolean doResolveAllCached(String hostname,
+    static ResolvedAddresses doResolveAllCached(String hostname,
                                       DnsRecord[] additionals,
-                                      Promise<List<InetAddress>> promise,
                                       DnsCache resolveCache,
                                       InternetProtocolFamily[] resolvedInternetProtocolFamilies) {
+        if (resolveCache == null) {
+            return null;
+        }
+
         final List<? extends DnsCacheEntry> cachedEntries = resolveCache.get(hostname, additionals);
         if (cachedEntries == null || cachedEntries.isEmpty()) {
-            return false;
-        }
-
-        Throwable cause = cachedEntries.get(0).cause();
-        if (cause == null) {
-            List<InetAddress> result = null;
-            final int numEntries = cachedEntries.size();
-            for (InternetProtocolFamily f : resolvedInternetProtocolFamilies) {
-                for (int i = 0; i < numEntries; i++) {
-                    final DnsCacheEntry e = cachedEntries.get(i);
-                    if (f.addressType().isInstance(e.address())) {
-                        if (result == null) {
-                            result = new ArrayList<InetAddress>(numEntries);
-                        }
-                        result.add(e.address());
+            return null;
+        }
+
+        DnsCacheEntry e0 = cachedEntries.get(0);
+        if (e0.cause() != null) {
+            return new ResolvedAddresses(e0.cause());
+        }
+
+        List<InetAddress> result = null;
+        for (InternetProtocolFamily f : resolvedInternetProtocolFamilies) {
+            for (DnsCacheEntry e : cachedEntries) {
+                if (f.addressType().isInstance(e.address())) {
+                    if (result == null) {
+                        result = new ArrayList<InetAddress>(cachedEntries.size());
                     }
+                    result.add(e.address());
                 }
             }
-            if (result != null) {
-                trySuccess(promise, result);
-                return true;
-            }
-            return false;
-        } else {
-            tryFailure(promise, cause);
-            return true;
         }
+        return result != null ? new ResolvedAddresses(result) : null;
```

Comment:
Isn't there a way to not induce extra allocations when there resolveCacheFallback is null?

Reply:
My first try (not visible in this PR) for this was to have this method return `Object`, which could either be null, an exception or a list. Returning `Object` requires the caller to do several `instanceof` checks and (generic) casts and handle the trySuccess/tryFailure logic. This method is called in several places, so returning `Object` would require duplication of this error-prone logic and makes the code more difficult to understand.

IMO, we shouldn't worry about creating such tiny and transient `ResolvedAddresses` objects. They will get almost immediately dereferenced and available for GC.